infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/StateTransfer.java

```diff
@@ -0,0 +1,117 @@
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.corfudb.protocols.wireprotocol.ILogData;
+import org.corfudb.protocols.wireprotocol.LogData;
+import org.corfudb.protocols.wireprotocol.Token;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.view.Layout;
+import org.corfudb.util.CFUtils;
+
+/**
+ * State transfer utility.
+ * Created by zlokhandwala on 2019-02-06.
+ */
+@Slf4j
+public class StateTransfer {
+
+    private StateTransfer() {
+        // Hide implicit public constructor.
+    }
+
+    /**
+     * Fetch and propagate the trimMark to the new/healing nodes. Else, a FastLoader reading from
+     * them will have to mark all the already trimmed entries as holes.
+     * Transfer an address segment from a cluster to a set of specified nodes.
+     * There are no cluster reconfigurations, hence no epoch change side effects.
+     *
+     * @param layout    layout
+     * @param endpoints destination nodes
+     * @param runtime   The runtime to read the segment from
+     * @param segment   segment to transfer
+     */
+    public static void transfer(Layout layout,
+                                Set<String> endpoints,
+                                CorfuRuntime runtime,
+                                Layout.LayoutSegment segment) throws ExecutionException, InterruptedException {
+
+        int batchSize = runtime.getParameters().getBulkReadSize();
+
+        long trimMark = runtime.getAddressSpaceView().getTrimMark().getSequence();
+        // Send the trimMark to the new/healing nodes.
+        // If this times out or fails, the Action performing the stateTransfer fails and retries.
+        for (String endpoint : endpoints) {
+            // TrimMark is the first address present on the log unit server.
+            // Perform the prefix trim on the preceding address = (trimMark - 1).
+            // Since the LU will reject trim decisions made from older epochs, we
+            // need to adjust the new trim mark to have the new layout's epoch.
+            Token prefixToken = new Token(layout.getEpoch(), trimMark - 1);
+            CFUtils.getUninterruptibly(runtime.getLayoutView().getRuntimeLayout(layout)
+                    .getLogUnitClient(endpoint)
+                    .prefixTrim(prefixToken));
+        }
+
+        if (trimMark > segment.getEnd()) {
+            log.info("stateTransfer: Nothing to transfer, trimMark {} greater than end of segment {}",
+                    trimMark, segment.getEnd());
+            return;
+        }
+
+        // State transfer should start from segment start address or trim mark whichever is lower.
+        long segmentStart = Math.max(trimMark, segment.getStart());
+
+        for (long chunkStart = segmentStart; chunkStart < segment.getEnd()
+                ; chunkStart = chunkStart + batchSize) {
+            long chunkEnd = Math.min((chunkStart + batchSize - 1), segment.getEnd() - 1);
+
+            long ts1 = System.currentTimeMillis();
```

Comment:
let's replace this by metrics, let's use metrics if we need to measure performance here

Reply:
Why only for datapath? We need metrics for everything. This is a new/updated functional, and we don't use metrics, we ignore it. I believe that's not right